libkbfs/file_data.go

```diff
@@ -97,6 +122,341 @@ func (fd *fileData) getFileBlockAtOffset(ctx context.Context,
 	return ptr, parentBlocks, block, nextBlockStartOff, startOff, wasDirty, nil
 }
 
+// getLeafBlocksForOffsetRange fetches all the leaf ("direct") blocks
+// that encompass the given offset range (half-inclusive) in the file.
+// If `endOff` is -1, it returns blocks until reaching the end of the
+// file.  Note the range could be made up of holes, indicated in the
+// IPtrs list of the parent blocks.  If `prefixOk` is true, the
+// function will ignore context deadline errors and return whatever
+// prefix of the data it could fetch within the deadine.  Return
+// params:
+//
+//   * pathsFromRoot is a slice, ordered by file offset, of paths from
+//     the root to each block that makes up the range.
+//   * blocks: a map from block pointer to a data-containing leaf node
+//     in the given range of offsets.
+//   * nextBlockOff is the offset of the block that follows the last
+//     block given in `pathsFromRoot`.  If `pathsFromRoot` contains
+//     the last block among the children, nextBlockOff is -1.
+func (fd *fileData) getLeafBlocksForOffsetRange(ctx context.Context,
+	pblock *FileBlock, startOff int64, endOff int64,
+	prefixOk bool) (pathsFromRoot [][]parentBlockAndChildIndex,
+	blocks map[BlockPointer]*FileBlock, nextBlockOffset int64, err error) {
+	if !pblock.IsInd {
+		// Return a single empty path and am empty child block map.
+		return [][]parentBlockAndChildIndex{nil}, nil, -1, nil
+	}
+
+	type resp struct {
+		pathsFromRoot [][]parentBlockAndChildIndex
+		blocks        map[BlockPointer]*FileBlock
+		nextBlockOff  int64
```

Comment:
I'd prefer `nextBlockOffset`, to be consistent with the name of the return param